File: binarySearch/fourSum2.py
import collections
class Solution:
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        count = 0
        dictionary = {}
        for i in A:
            for j in B:
                val = i+j
                if val in dictionary:
                    dictionary[val]+=1
                else:
                    dictionary[val] = 1
                    
        for i in range(len(C)):
            for j in range(len(D)):
                target = 0 - (C[i]+D[j])
                if target in dictionary:
                    count += dictionary[target]
        return count

File: binarySearch/fourSum2.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        unordered_map<int, int> count;
        int sum1 = 0, counter = 0;
        for (int a: A){
            for (int b: B){
                count[a+b]++;    
            }
        }
        
        for (int c: C){
            for(int d: D){
                int target = 0 - (c+d);
                if (count.count(target))
                    counter+=count[target];
            }
        }
        return counter;
    }
};
